trunk/NDHMS/Routing/Reservoirs/module_read_timeslice_data.F

```diff
@@ -0,0 +1,354 @@
+! This module reads USGS timeslice files to get gage discharge
+! values that will be used by reservoirs. An observation lookback
+! period is passed in to determine how far back in time from the
+! current model time the module will look for timeslice files.
+! The observation resolution determines the time increments the
+! module will look back. For instance, a standard lookback period
+! would be 18 hours with an observation resolution of 15 minutes,
+! where a model current time of 8:00 PM would search for timeslice
+! files at every 15 minute increment between 2:00 AM and 8:00 PM
+! that day. The module will first search for the most recent
+! timeslice files and grab the discharge for a particular
+! lake/reservoir if the gage quality standard is met at that time.
+! If a gage discharge is missing or if the gage quality standard
+! is not met for any particular lake/reservoir in the given
+! timeslice file, the module will continue to look back at every
+! observation resolution increment until either all
+! lakes/reservoirs have a good quality discharge or the end of
+! the lookback period is reached. The total lookback seconds
+! from current model time that the discharge is read will also
+! be returned.
+
+#define days_in_year 365
```

Comment:
This is not good practice. These should be declared as module variables.

Reply:
Agreed they should be something like

```
module module_read_timeslice_data
   ...
   integer, parameter :: days_in_year = 365

```

Reply:
Mind elaborating on why this isn't good practice?  I'm just curious as to the rational.  Constants as macros can be effective and efficient, and for something that isn't "module" specific (i.e. days_in_year is needed by the module, but doesn't define the module and/or its capability...), they can be shared via a common definition header.

Reply:
To borrow a phrase from the Python people, it isn't Fortranic. If it's defined as a constant component with `parameter` attribute, the compiler will inline it efficiently. These can certainly be put in a common constant module, imported via `use, only:` syntax where needed.

Reply:
It's not actually Fortran, let alone "Fortranic". Imagine if I were proposing including floats into the code via sed. It's not wildly different. I could call sed in the make files...  CPP sneaks in because it has other uses, but that's just not the way to do it. We should have a constants module (or object, though it's data would apparently all be public). I thought there was an issue about this but I dont see one. Then constants would be available for all needed types via that module. That would eliminate lots of magic numbers (including ones, zeros, etc). It would also focus on what our standard defaults are (false_int maybe?). I would put this forward as a low-hanging fruit along with a time-module/object for the next year of refactoring.

CPP has been employed  to provide really high-level compile-time structures. In general, we should remove all of that. For example CESM has moved away from all CPP and compile time options. I still think we'd have major build compiltime options, but not via CPP.

Reply:
I generally agree about parameters/magic numbers.  The one benefit of using the pre-processor for these is actually for type safety, but that is minor gain.  For other things, such as defining logging levels/debugging/compiler specific things, I think CPP is worth keeping around.  Anything platform/architecture specific is well managed by build systems to set these and build the appropriate configuration.  Care does need to be taken in keeping code readable and organized with the macros though.

Reply:
@jdmattern-noaa fixing. Removing days_in_year because no longer needed. Changing others to parameter attribute in the module.